src/errors.rs

```diff
@@ -1,31 +1,82 @@
+//            BoxcarError::AttributeError(attribute_error) => write!(f, "attribute error: {}", attribute_error),
+//        }
+//    }
+//}
+//
+//impl Error for BoxcarError {
+//    fn source(&self) -> Option<&(dyn Error + 'static)> {
+//        match self {
+//            BoxcarError::ParseError(_, _, parse_error) => Some(parse_error),
+//            _ => None
+//        }
+//    }
+//}
+
+
+#[derive(PartialEq, Debug, Clone)]
 pub enum ParseError {
-    #[fail(display = "A size of zero is not valid")]
+    ParseError(&'static str, i32, Box<ParseError>),
     ZeroSize,
-
-    #[fail(display = "Unable decode data as utf8: {}", _0)]
-    Utf8Error(#[cause] str::Utf8Error),
-
-    #[fail(display = "Text of size {} is too large", _0)]
+    Utf8Error(str::Utf8Error),
     TextTooLarge(i32),
-
-    #[fail(
-        display = "Insufficient data. Expected {} bytes, but only {} left",
-        _0, _1
-    )]
     InsufficientData(i32, i32),
-
-    #[fail(display = "Did not expect a property of: {}", _0)]
     UnexpectedProperty(String),
-
-    #[fail(display = "Crc mismatch. Expected {} but received {}", _0, _1)]
     CrcMismatch(u32, u32),
-
-    #[fail(display = "list of size {} is too large", _0)]
+    CorruptReplay(String, Box<ParseError>),
     ListTooLarge(usize),
+    NetworkError(NetworkError)
+}
+
+impl Display for ParseError{
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            ParseError::ZeroSize => write!(f, "A size of zero is not valid"),
+            ParseError::Utf8Error(utf8_error) => write!(f, "Unable decode data as utf8: {}", utf8_error),
+            ParseError::TextTooLarge(size) => write!(f, "Text of size {} is too large", size),
+            ParseError::InsufficientData(expected, left) => write!(f, "Insufficient data. Expected {} bytes, but only {} left", expected, left),
+            ParseError::UnexpectedProperty(property) => write!(f, "Did not expect a property of: {}", property),
+            ParseError::CrcMismatch(expected, found) => write!(f, "Crc mismatch. Expected {} but received {}", expected, found),
+            ParseError::CorruptReplay(section, _) => write!(f, "Failed to parse {} and crc check failed. Replay is corrupt", section),
+            ParseError::ListTooLarge(size) => write!(f, "list of size {} is too large", size),
+            ParseError::ParseError(section, bytes_read, parse_error) => write!(f, "Could not decode replay {} at offset ({}): {}", section, bytes_read, parse_error),
+            ParseError::NetworkError(network_error) => write!(f, "{}", network_error)
+        }
+    }
+}
+
+impl Error for ParseError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            ParseError::Utf8Error(utf8_error) => Some(utf8_error),
+            ParseError::CorruptReplay(_, error) => Some(error),
+            ParseError::ParseError(_, _, error) => Some(error),
```

Comment:
Does network error need to be included in this list?

Reply:
I switched the return type of most methods to `Result<T, ParseError>`. I thought it made since I wouldn't expect to see a method named `parse()` to return a `NetworkError`. It seems like the class `NetworkError` is still describing a failing to parse the network section of the replay so I felt ok to embed the `Error` in `ParseError`

Reply:
I think we may be talking about two different things. I had adding `ParseError::NetworkError` branch in mind -- you shouldn't need to change any other code.

```rust
impl Error for ParseError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            ParseError::Utf8Error(utf8_error) => Some(utf8_error),
            ParseError::CorruptReplay(_, error) => Some(error),
            ParseError::ParseError(_, _, error) => Some(error),
            ParseError::NetworkError(error) => Some(error),
            _ => None
        }
    }
}
```

Reply:
Ah yes. Just because it was the source of the error. Made sense to return it there.

Reply:
I'm not that attached to it. I can remove it if you like.

Reply:
AFAIK, you don't have to remove anything, this is the diff I'm suggesting.

```diff
 impl Error for ParseError {
     fn source(&self) -> Option<&(dyn Error + 'static)> {
         match self {
             ParseError::Utf8Error(utf8_error) => Some(utf8_error),
             ParseError::CorruptReplay(_, error) => Some(error),
             ParseError::ParseError(_, _, error) => Some(error),
+            ParseError::NetworkError(error) => Some(error),
             _ => None
         }
     }
 }
```

Reply:
Oops! I thought this was already the case! Good catch.